### DISCUSSION TITLE
No Preview Available using S3 storage

### DISCUSSION THREAD

**Discussion by stevolucidity** [pixelfed/pixelfed#5143] [Support]
I've managed to get _some_ objects stored in my Wasabi bucket. I can see the new folders being stored in the Wasabi control panel.

However, when I 'switch on' S3 storage I start to get the 'No Preview Available' problem.

The S3 section in my `.env`:
````
PF_ENABLE_CLOUD=true
FILESYSTEM_DRIVER=local
FILESYSTEM_CLOUD=s3
AWS_ACCESS_KEY_ID=[REDACTED]
AWS_SECRET_ACCESS_KEY=[REDACTED]
AWS_DEFAULT_REGION=eu-west-1
AWS_BUCKET=twd-pixelfed-media
AWS_URL=https://media.thewalkingdeaf.net
AWS_ENDPOINT=https://s3.eu-west-1.wasabisys.com
#AWS_USE_PATH_STYLE_ENDPOINT=false
````

The bucket policy:
````
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "AddPerm",
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": "s3:GetObject",
      "Resource": "arn:aws:s3:::twd-pixelfed-media/*"
    }
  ]
}
````

The bucket contents:

![15ED52EB-6D27-4E38-8F30-DACCB4C5F3B2](https://user-images.githubusercontent.com/44602820/202849078-d7857775-9232-42cc-84b1-b617c1614ed9.jpeg)

I'm so close to getting it working I can taste it!
Can someone please help me resolve this?

Attached image shows home feed just after 'switching on' S3 storage again.

![A80481D3-77D7-403B-8DB3-6E33DD90D76D](https://user-images.githubusercontent.com/44602820/202848709-9c5dce02-15fb-483d-8d70-1c1f61ad2352.jpeg)

**Comment by stevolucidity** [comment]
To clarify, nothing is showing in the nginx, pixelfed or laravel logs to indicate an issue.

**Comment by ghost** [comment]
You're definitely closer than I was with implementing Wasabi's S3 storage to my instance. Don't know if this works, but I do also run my own Mastodon-instance without any issues and the policy for this looks like the following:

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "s3:ListBucket",
      "Resource": "arn:aws:s3:::pixelfed-media"
    },
    {
      "Effect": "Allow",
      "Action": [
        "s3:PutObject",
        "s3:GetObject",
        "s3:DeleteObject"
      ],
      "Resource": "arn:aws:s3:::pixelfed-media/*"
    }
  ]
}
```

Could you try this example to see whether the preview can be generated now?

Don't know if this works but I feel the same - I was also so close that I thought I was just the glimpse of an eye away but didn't succeed so far ;) !

Best regards,

Oliver

**Comment by stevolucidity** [comment]
Ah thanks, @OliverPifferi but I've even tried with s3:* as the action with no success. 

If I look at a global feed or indeed my own feed, I can see from the image link where PixelFed 'thinks' the image should be, but the location doesn't exist on the remote storage. This is also true for the AWS S3 I tried, so it's not peculiar to wasabi. 

Without an error conditions, there's nothing logged to let me follow the attempt the write. The database must record the alleged location, though. It's just that the write must fail, but isn't logged. 

I have both mastodon and peertube instances and it's working fine for both of them. 

I've given up banging my head against that particular wall for now. I imagine the idea will gain traction in the near future and someone smarter and more important than me will either find and post the answer or at least get a fix for it. 

Best of luck with your endeavours!

**Comment by ghost** [comment]
> Ah thanks, @OliverPifferi but I've even tried with s3:* as the action with no success.
> 
> If I look at a global feed or indeed my own feed, I can see from the image link where PixelFed 'thinks' the image should be, but the location doesn't exist on the remote storage. This is also true for the AWS S3 I tried, so it's not peculiar to wasabi.
> 
> Without an error conditions, there's nothing logged to let me follow the attempt the write. The database must record the alleged location, though. It's just that the write must fail, but isn't logged.
> 
> I have both mastodon and peertube instances and it's working fine for both of them.
> 
> I've given up banging my head against that particular wall for now. I imagine the idea will gain traction in the near future and someone smarter and more important than me will either find and post the answer or at least get a fix for it.
> 
> Best of luck with your endeavours!

Thanks for the exchange here... When you use Mastodon and Peertube with a similar config (I do with Mastodon), it must indeed be something different. What a bummer - still waiting for a solution here so that I can set up my own Pixelfed-instance with S3-storage.

Don't think it has something to do with being smart - as you seem to be caching the S3 via https://media.thewalkingdeaf.net, maybe it could be related to the ghost-config in NGINX? Had a similar case with a typo and Mastodon once - but this is just a desperate assumption.

Holding thumbs, let me know if it worked somehow!

BR,

Oliver

**Comment by okpierre** [comment]
Try regenerate thumbnails @evoLucidity @OliverPifferi 

URLs might be hard coded - any mysql db expert can confirm?

If you have lots of image then you might run into this issue https://github.com/pixelfed/pixelfed/issues/3795

**Comment by mitexleo** [comment]
Facing same issue with Storj DCS.

**Comment by ghost** [comment]
> Try regenerate thumbnails @evoLucidity @OliverPifferi
> 
> URLs might be hard coded - any mysql db expert can confirm?
> 
> If you have lots of image then you might run into this issue #3795

This helped to see thumbnails of the uploaded files in the Admin-section. Tried this with a new instance and one pic - same effect. In Wasabi, I can browse and download the files so it must be any kind of the way how Pixelfed handles the uploaded pictures.

I also the the correct thumbnail in my profile but not on the main screen - seems to be very, very close to the solution.

<img width="767" alt="Bildschirm­foto 2022-11-26 um 13 24 27" src="https://user-images.githubusercontent.com/42076599/204088825-3b0a6384-cde8-4a03-8077-6f424266c73f.png">
<img width="688" alt="Bildschirm­foto 2022-11-26 um 13 24 33" src="https://user-images.githubusercontent.com/42076599/204088828-3d38a35d-6a96-4c21-8f01-6f337d09c889.png">

**Comment by ghost** [comment]
Got it... I removed-/disabled the individual "AWS_URL"-parameter in the .env while using Wasabi:

I can confirm that the previews are now being generated as expected while serving via S3. Can anyone of the other folks here @evoLucidity @mitexleo test and confirm this, too?

## S3 Configuration (Post-Installer)
PF_ENABLE_CLOUD=true
FILESYSTEM_DRIVER=local
FILESYSTEM_CLOUD=s3
AWS_ACCESS_KEY_ID=[REDACTED]
AWS_SECRET_ACCESS_KEY=[REDACTED]
AWS_DEFAULT_REGION=eu-central-2
AWS_BUCKET=your-pixelfed-s3-bucket
#AWS_URL=uncomment_this
AWS_ENDPOINT=https://s3.eu-central-2.wasabisys.com
#AWS_USE_PATH_STYLE_ENDPOINT=false

**Comment by stevolucidity** [comment]
Ooooh! Will try this when I get home shortly! 🤓

Seem to remember having tried this, but will test soon and report back.

**Comment by stevolucidity** [comment]
Okay, so this work so long as I specify the actual endpoint as you do above. 🎉

However I was really wanting to reverse-proxy it through my nginx server like I do with mastodon and peertube.
I know, I want the moon on a stick!

When I try this I get CURL error 60 and I'm not really sure why. I'll keep trying...

**Comment by ghost** [comment]
At least it can be reproduced and albeit I would also prefer a reverse-proxying like I did it with Mastodon, too, it's at least a step in the proper direction.

I've got it running for myself and although caching it through nginx would be the peak of perfection, a bird in the hand is worth two in the bush in my use case.

Maybe we discover the last one, two percents in the near future 😉 !

**Comment by stevolucidity** [comment]
I agree. I'm happy to use it as it is for now. I'm thinking we should be able to remove the local `storage/app/public` folder after an `aws s3 sync`?
I'll rename it first to be on the safe side.

For anyone follow this, I can confirm that the `.env` snippet by @OliverPifferi above works.
I can also confirm that adding the `AWS_URL` line back in to attempt to cache through local nginx reverse proxy re-introduces the `No Preview Available` issue.

Let's hope we stumble across as caching solution soon.

**Comment by mitexleo** [comment]
> Got it... I removed-/disabled the individual "AWS_URL"-parameter in the .env while using Wasabi:
> 
> I can confirm that the previews are now being generated as expected while serving via S3. Can anyone of the other folks here @evoLucidity @mitexleo test and confirm this, too?
> 
> ## S3 Configuration (Post-Installer)
> PF_ENABLE_CLOUD=true FILESYSTEM_DRIVER=local FILESYSTEM_CLOUD=s3 AWS_ACCESS_KEY_ID=[REDACTED] AWS_SECRET_ACCESS_KEY=[REDACTED] AWS_DEFAULT_REGION=eu-central-2 AWS_BUCKET=your-pixelfed-s3-bucket #AWS_URL=uncomment_this AWS_ENDPOINT=https://s3.eu-central-2.wasabisys.com #AWS_USE_PATH_STYLE_ENDPOINT=false

Yep ..  That was the cause... But In case of Storj DCS I had to setup GatewayST ... I'm using AWS S3 now ...

**Comment by ehsan0921** [comment]
> > Got it... I removed-/disabled the individual "AWS_URL"-parameter in the .env while using Wasabi:
> > I can confirm that the previews are now being generated as expected while serving via S3. Can anyone of the other folks here @evoLucidity @mitexleo test and confirm this, too?
> > ## S3 Configuration (Post-Installer)
> > PF_ENABLE_CLOUD=true FILESYSTEM_DRIVER=local FILESYSTEM_CLOUD=s3 AWS_ACCESS_KEY_ID=[REDACTED] AWS_SECRET_ACCESS_KEY=[REDACTED] AWS_DEFAULT_REGION=eu-central-2 AWS_BUCKET=your-pixelfed-s3-bucket #AWS_URL=uncomment_this AWS_ENDPOINT=https://s3.eu-central-2.wasabisys.com #AWS_USE_PATH_STYLE_ENDPOINT=false
> 
> Yep .. That was the cause... But In case of Storj DCS I had to setup GatewayST ... I'm using AWS S3 now ...

Hey mate
can you please help with the S3 or DO.
I cant set it,
it gose local.
plesae send me a sample of S3 or DO please.

**Comment by mitexleo** [comment]
Dl

> > > Got it... I removed-/disabled the individual "AWS_URL"-parameter in the .env while using Wasabi:
> > > I can confirm that the previews are now being generated as expected while serving via S3. Can anyone of the other folks here @evoLucidity @mitexleo test and confirm this, too?
> > > ## S3 Configuration (Post-Installer)
> > > PF_ENABLE_CLOUD=true FILESYSTEM_DRIVER=local FILESYSTEM_CLOUD=s3 AWS_ACCESS_KEY_ID=[REDACTED] AWS_SECRET_ACCESS_KEY=[REDACTED] AWS_DEFAULT_REGION=eu-central-2 AWS_BUCKET=your-pixelfed-s3-bucket #AWS_URL=uncomment_this AWS_ENDPOINT=https://s3.eu-central-2.wasabisys.com #AWS_USE_PATH_STYLE_ENDPOINT=false
> > 
> > 
> > Yep .. That was the cause... But In case of Storj DCS I had to setup GatewayST ... I'm using AWS S3 now ...
> 
> Hey mate can you please help with the S3 or DO. I cant set it, it gose local. plesae send me a sample of S3 or DO please.

Did you set FILESYSTEM_DRIVER to s3?

**Comment by ghost** [comment]
> > > Got it... I removed-/disabled the individual "AWS_URL"-parameter in the .env while using Wasabi:
> 
> > > I can confirm that the previews are now being generated as expected while serving via S3. Can anyone of the other folks here @evoLucidity @mitexleo test and confirm this, too?
> 
> > > ## S3 Configuration (Post-Installer)
> 
> > > PF_ENABLE_CLOUD=true FILESYSTEM_DRIVER=local FILESYSTEM_CLOUD=s3 AWS_ACCESS_KEY_ID=[REDACTED] AWS_SECRET_ACCESS_KEY=[REDACTED] AWS_DEFAULT_REGION=eu-central-2 AWS_BUCKET=your-pixelfed-s3-bucket #AWS_URL=uncomment_this AWS_ENDPOINT=https://s3.eu-central-2.wasabisys.com #AWS_USE_PATH_STYLE_ENDPOINT=false
> 
> > 
> 
> > Yep .. That was the cause... But In case of Storj DCS I had to setup GatewayST ... I'm using AWS S3 now ...
> 
> 
> 
> Hey mate
> 
> can you please help with the S3 or DO.
> 
> I cant set it,
> 
> it gose local.
> 
> plesae send me a sample of S3 or DO please.
> 
> 
> 
> 

The example I gave with filesystem driver being "local" exactly works in my environment, though profile pictures are stored locally while all media is uploaded to S3. Still need to verify this for my use case.

**Comment by okpierre** [comment]
@mitexleo during my test, couldn't get the settings `filesystems_driver="local"` as @OliverPifferi pointed out to work 

All urls shown in browser point to local storage only but with `filesystems_driver="s3"` it points to s3 urls

Besides env targeted at "spaces" anyone know if there is underlying differences between https://github.com/pixelfed/pixelfed/blob/4393780a4d3ff2ff30d26ec126fc2afc668fa420/config/filesystems.php#L59 and https://github.com/pixelfed/pixelfed/blob/4393780a4d3ff2ff30d26ec126fc2afc668fa420/config/filesystems.php#L72

Now it leads me to wonder if there is something else that is causing these differences https://github.com/pixelfed/pixelfed/issues/3559 https://github.com/pixelfed/pixelfed/issues/3832 https://github.com/pixelfed/pixelfed/issues/3775 https://github.com/pixelfed/pixelfed/issues/2866

**Comment by ghost** [comment]
Running my test instance (without a caching proxy so far - this is still the next step), implemented from the scratch, now completely via S3 by using

`
## S3 Configuration (Post-Installer)
PF_ENABLE_CLOUD=true
FILESYSTEM_DRIVER=s3
FILESYSTEM_CLOUD=s3
AWS_ACCESS_KEY_ID=[REDACTED]
AWS_SECRET_ACCESS_KEY=[REDACTED]
AWS_DEFAULT_REGION=eu-central-2
AWS_BUCKET=pixelfed
AWS_ENDPOINT=https://s3.eu-central-2.wasabisys.com
#AWS_USE_PATH_STYLE_ENDPOINT=false`

**Comment by mitexleo** [comment]
> Got it... I removed-/disabled the individual "AWS_URL"-parameter in the .env while using Wasabi:
> 
> I can confirm that the previews are now being generated as expected while serving via S3. Can anyone of the other folks here @evoLucidity @mitexleo test and confirm this, too?
> 
> ## S3 Configuration (Post-Installer)
> PF_ENABLE_CLOUD=true FILESYSTEM_DRIVER=local FILESYSTEM_CLOUD=s3 AWS_ACCESS_KEY_ID=[REDACTED] AWS_SECRET_ACCESS_KEY=[REDACTED] AWS_DEFAULT_REGION=eu-central-2 AWS_BUCKET=your-pixelfed-s3-bucket #AWS_URL=uncomment_this AWS_ENDPOINT=https://s3.eu-central-2.wasabisys.com #AWS_USE_PATH_STYLE_ENDPOINT=false

Yep ... That works 💪  ... Thanks Mate 😊

**Comment by dansup** [comment]
Remove `AWS_URL` then run `php artisan config:cache && php artisan cache:clear`

**Comment by dansup** [comment]
**Do not set `FILESYSTEM_DRIVER=s3`**

Keep it set to `FILESYSTEM_DRIVER=local` and set `FILESYSTEM_CLOUD=s3` or it will break things

**Comment by mitexleo** [comment]
> **Do not set `FILESYSTEM_DRIVER=s3`**
> 
> Keep it set to `FILESYSTEM_DRIVER=local` and set `FILESYSTEM_CLOUD=s3` or it will break things

😘

**Comment by stevolucidity** [comment]
Thanks for the confirmation that we're on the right track, @dansup! 😀
Have you been able to cache through a proxy successfully?

**Comment by ghost** [comment]
> **Do not set `FILESYSTEM_DRIVER=s3`**
> 
> 
> 
> Keep it set to `FILESYSTEM_DRIVER=local` and set `FILESYSTEM_CLOUD=s3` or it will break things

Thanks for clarifying! 👍🏻

**Comment by amozoss** [comment]
> Facing same issue with Storj DCS.

Works for me on Storj when I removed `AWS_URL` and `AWS_USE_PATH_STYLE_ENDPOINT=false` 

```
## S3 Configuration (Post-Installer)
PF_ENABLE_CLOUD=true
FILESYSTEM_DRIVER=local
FILESYSTEM_CLOUD=s3
AWS_ACCESS_KEY_ID=<redacted>
AWS_SECRET_ACCESS_KEY=<redacted>
AWS_DEFAULT_REGION=us1
AWS_BUCKET=pixelfed-test
#AWS_URL=
AWS_ENDPOINT=https://gateway.storjshare.io
#AWS_USE_PATH_STYLE_ENDPOINT=false
```

**Comment by ehsan0921** [comment]
how to get AWS endpoint in S3 Amazon?

**Comment by ehsan0921** [comment]
> > Facing same issue with Storj DCS.
> 
> Works for me on Storj when I removed `AWS_URL` and `AWS_USE_PATH_STYLE_ENDPOINT=false`
> 
> ```
> ## S3 Configuration (Post-Installer)
> PF_ENABLE_CLOUD=true
> FILESYSTEM_DRIVER=local
> FILESYSTEM_CLOUD=s3
> AWS_ACCESS_KEY_ID=<redacted>
> AWS_SECRET_ACCESS_KEY=<redacted>
> AWS_DEFAULT_REGION=us1
> AWS_BUCKET=pixelfed-test
> #AWS_URL=
> AWS_ENDPOINT=https://gateway.storjshare.io
> #AWS_USE_PATH_STYLE_ENDPOINT=false
> ```

Also I need to mention it works but it will upload on local always. not on cloud. check you images path.

**Comment by amozoss** [comment]
> > > Facing same issue with Storj DCS.
> > 
> > 
> > Works for me on Storj when I removed `AWS_URL` and `AWS_USE_PATH_STYLE_ENDPOINT=false`
> > ```
> > ## S3 Configuration (Post-Installer)
> > PF_ENABLE_CLOUD=true
> > FILESYSTEM_DRIVER=local
> > FILESYSTEM_CLOUD=s3
> > AWS_ACCESS_KEY_ID=<redacted>
> > AWS_SECRET_ACCESS_KEY=<redacted>
> > AWS_DEFAULT_REGION=us1
> > AWS_BUCKET=pixelfed-test
> > #AWS_URL=
> > AWS_ENDPOINT=https://gateway.storjshare.io
> > #AWS_USE_PATH_STYLE_ENDPOINT=false
> > ```
> 
> Also I need to mention it works but it will upload on local always. not on cloud. check you images path.

Ah indeed, didn't notice that it sometimes shows the local and sometimes shows the cloud in the feed. I'm looking into a way to change which URL it uses. If that's possible I think I can get it to work

**Comment by mitexleo** [comment]
> > > > Facing same issue with Storj DCS.
> > > 
> > > 
> > > Works for me on Storj when I removed `AWS_URL` and `AWS_USE_PATH_STYLE_ENDPOINT=false`
> > > ```
> > > ## S3 Configuration (Post-Installer)
> > > PF_ENABLE_CLOUD=true
> > > FILESYSTEM_DRIVER=local
> > > FILESYSTEM_CLOUD=s3
> > > AWS_ACCESS_KEY_ID=<redacted>
> > > AWS_SECRET_ACCESS_KEY=<redacted>
> > > AWS_DEFAULT_REGION=us1
> > > AWS_BUCKET=pixelfed-test
> > > #AWS_URL=
> > > AWS_ENDPOINT=https://gateway.storjshare.io
> > > #AWS_USE_PATH_STYLE_ENDPOINT=false
> > > ```
> > 
> > 
> > Also I need to mention it works but it will upload on local always. not on cloud. check you images path.
> 
> Ah indeed, didn't notice that it sometimes shows the local and sometimes shows the cloud in the feed. I'm looking into a way to change which URL it uses. If that's possible I think I can get it to work

Actually,  The file will be first uploaded to local ... After processing pixelfed will upload the file to S3 .. and the URL wil be updated after a few minutes

**Comment by mitexleo** [comment]
> > Facing same issue with Storj DCS.
> 
> Works for me on Storj when I removed `AWS_URL` and `AWS_USE_PATH_STYLE_ENDPOINT=false`
> 
> ```
> ## S3 Configuration (Post-Installer)
> PF_ENABLE_CLOUD=true
> FILESYSTEM_DRIVER=local
> FILESYSTEM_CLOUD=s3
> AWS_ACCESS_KEY_ID=<redacted>
> AWS_SECRET_ACCESS_KEY=<redacted>
> AWS_DEFAULT_REGION=us1
> AWS_BUCKET=pixelfed-test
> #AWS_URL=
> AWS_ENDPOINT=https://gateway.storjshare.io
> #AWS_USE_PATH_STYLE_ENDPOINT=false
> ```

Did you hosted GatewayST ?

**Comment by ehsan0921** [comment]
> > > > > Facing same issue with Storj DCS.
> > > > 
> > > > 
> > > > Works for me on Storj when I removed `AWS_URL` and `AWS_USE_PATH_STYLE_ENDPOINT=false`
> > > > ```
> > > > ## S3 Configuration (Post-Installer)
> > > > PF_ENABLE_CLOUD=true
> > > > FILESYSTEM_DRIVER=local
> > > > FILESYSTEM_CLOUD=s3
> > > > AWS_ACCESS_KEY_ID=<redacted>
> > > > AWS_SECRET_ACCESS_KEY=<redacted>
> > > > AWS_DEFAULT_REGION=us1
> > > > AWS_BUCKET=pixelfed-test
> > > > #AWS_URL=
> > > > AWS_ENDPOINT=https://gateway.storjshare.io
> > > > #AWS_USE_PATH_STYLE_ENDPOINT=false
> > > > ```
> > > 
> > > 
> > > Also I need to mention it works but it will upload on local always. not on cloud. check you images path.
> > 
> > 
> > Ah indeed, didn't notice that it sometimes shows the local and sometimes shows the cloud in the feed. I'm looking into a way to change which URL it uses. If that's possible I think I can get it to work
> 
> Actually, The file will be first uploaded to local ... After processing pixelfed will upload the file to S3 .. and the URL wil be updated after a few minutes

wow, Yes, you are right, 
thanks for letting me know.

**Comment by amozoss** [comment]
> > > Facing same issue with Storj DCS.
> > 
> > 
> > Works for me on Storj when I removed `AWS_URL` and `AWS_USE_PATH_STYLE_ENDPOINT=false`
> > ```
> > ## S3 Configuration (Post-Installer)
> > PF_ENABLE_CLOUD=true
> > FILESYSTEM_DRIVER=local
> > FILESYSTEM_CLOUD=s3
> > AWS_ACCESS_KEY_ID=<redacted>
> > AWS_SECRET_ACCESS_KEY=<redacted>
> > AWS_DEFAULT_REGION=us1
> > AWS_BUCKET=pixelfed-test
> > #AWS_URL=
> > AWS_ENDPOINT=https://gateway.storjshare.io
> > #AWS_USE_PATH_STYLE_ENDPOINT=false
> > ```
> 
> Did you hosted GatewayST ?

I actually used the [Storj hosted S3 compatible gateway](https://docs.storj.io/dcs/getting-started/gateway-mt). 

Here's the main things I had to do:
1. [Generate an access grant](https://docs.storj.io/dcs/getting-started/quickstart-uplink-cli/uploading-your-first-object/create-first-access-grant) and configure uplink
```bash
uplink setup
```

2. Make the bucket public (make sure your encryption passphrase for your S3 Creds are the same as your access grant used for uplink). 
```bash
uplink share --url --disallow-lists --not-after=none sj://pixelfed
...
=========== BROWSER URL ==================================================================
URL       : https://link.storjshare.io/s/jushfpcztig5wjpjovr3sztvz4fq/pixelfed
```
3. In the output browser URL, change `/s/` to `/raw/` and use it for AWS_URL.
Also set `PF_ENABLE_CLOUD=true` and `FILESYSTEM_DRIVER=s3`
```
## S3 Configuration (Post-Installer)
PF_ENABLE_CLOUD=true
FILESYSTEM_DRIVER=s3
FILESYSTEM_CLOUD=s3
AWS_ACCESS_KEY_ID=<redacted>
AWS_SECRET_ACCESS_KEY=<redacted>
AWS_DEFAULT_REGION=us1
AWS_BUCKET=pixelfed
AWS_URL=https://link.storjshare.io/raw/jushfpcztig5wjpjovr3sztvz4fq/pixelfed
AWS_ENDPOINT=https://gateway.storjshare.io
#AWS_USE_PATH_STYLE_ENDPOINT=false
```
4. reload config and restart services
```
sudo php artisan config:cache
sudo systemctl restart pixelfed nginx php8.1-fpm.service
```
Also here's a live stream I did with the Chief Architect of Storj, setting up pixelfed on ubuntu connected to storj  https://www.youtube.com/watch?v=K-Zmgt6KtzI&t=2307s

**Comment by mitexleo** [comment]
> > > > Facing same issue with Storj DCS.
> > > 
> > > 
> > > Works for me on Storj when I removed `AWS_URL` and `AWS_USE_PATH_STYLE_ENDPOINT=false`
> > > ```
> > > ## S3 Configuration (Post-Installer)
> > > PF_ENABLE_CLOUD=true
> > > FILESYSTEM_DRIVER=local
> > > FILESYSTEM_CLOUD=s3
> > > AWS_ACCESS_KEY_ID=<redacted>
> > > AWS_SECRET_ACCESS_KEY=<redacted>
> > > AWS_DEFAULT_REGION=us1
> > > AWS_BUCKET=pixelfed-test
> > > #AWS_URL=
> > > AWS_ENDPOINT=https://gateway.storjshare.io
> > > #AWS_USE_PATH_STYLE_ENDPOINT=false
> > > ```
> > 
> > 
> > Did you hosted GatewayST ?
> 
> I actually used the [Storj hosted S3 compatible gateway](https://docs.storj.io/dcs/getting-started/gateway-mt).
> 
> Here's the main things I had to do:
> 
> 1. [Generate an access grant](https://docs.storj.io/dcs/getting-started/quickstart-uplink-cli/uploading-your-first-object/create-first-access-grant) and configure uplink
> 
> ```shell
> uplink setup
> ```
> 
> 2. Make the bucket public (make sure your encryption passphrase for your S3 Creds are the same as your access grant used for uplink).
> 
> ```shell
> uplink share --url --disallow-lists --not-after=none sj://pixelfed
> ...
> =========== BROWSER URL ==================================================================
> URL       : https://link.storjshare.io/s/jushfpcztig5wjpjovr3sztvz4fq/pixelfed
> ```
> 
> 3. In the output browser URL, change `/s/` to `/raw/` and use it for AWS_URL.
>    Also set `PF_ENABLE_CLOUD=true` and `FILESYSTEM_DRIVER=s3`
> 
> ```
> ## S3 Configuration (Post-Installer)
> PF_ENABLE_CLOUD=true
> FILESYSTEM_DRIVER=s3
> FILESYSTEM_CLOUD=s3
> AWS_ACCESS_KEY_ID=<redacted>
> AWS_SECRET_ACCESS_KEY=<redacted>
> AWS_DEFAULT_REGION=us1
> AWS_BUCKET=pixelfed
> AWS_URL=https://link.storjshare.io/raw/jushfpcztig5wjpjovr3sztvz4fq/pixelfed
> AWS_ENDPOINT=https://gateway.storjshare.io
> #AWS_USE_PATH_STYLE_ENDPOINT=false
> ```
> 
> 4. reload config and restart services
> 
> ```
> sudo php artisan config:cache
> sudo systemctl restart pixelfed nginx php8.1-fpm.service
> ```
> 
> Also here's a live stream I did with the Chief Architect of Storj, setting up pixelfed on ubuntu connected to storj https://www.youtube.com/watch?v=K-Zmgt6KtzI&t=2307s

Wow 👌... But setting FILESYSTEM_DRIVER to s3 will cause many problem ...

**Comment by lapineige** [comment]
I have a similar issue, if not the same.

I am trying to setup Garage has a self-hosted S3 storage.

Here is my configuration :
```
PF_ENABLE_CLOUD=true
FILESYSTEM_CLOUD=s3
AWS_ACCESS_KEY_ID=SomeID
AWS_SECRET_ACCESS_KEY=SomeKey
AWS_DEFAULT_REGION=SomeRegion
AWS_BUCKET=pixelfed-media
#AWS_URL=
AWS_ENDPOINT=https://domain.tdl:443 # Because Nginx reverse-proxy
AWS_USE_PATH_STYLE_ENDPOINT=true
```
`FILESYSTEM_DRIVER` is *not* set (as it should).

I also have (had ?) this issue with `PF_ENABLE_CLOUD` https://github.com/pixelfed/pixelfed/issues/4899

---

New media upload fails with no preview available. The no preview image is shown, with a local URL.
In the Admin Media page, there I can see the broken image icon, and an URL for the original image, such as `https://garage.domain.tld/pixelfed-bucket/public/m/_v2/1/randomchain-randomchain/randomchain/longchain.jpg`.
If I reach this URL, I have a message telling me that anonymous access is not allowed. Is that the issue ?

Regenerating thumbnails didn't work. Worse: some old image thumbnails *and final picture* where no longer working. The URL where in the S3 storage and no longer local to the Pixelfed instance.

**Comment by lapineige** [comment]
> Regenerating thumbnails didn't work. Worse: some old image thumbnails and final picture where no longer working. The URL where in the S3 storage and no longer local to the Pixelfed instance.

You want a funny story ?
I disabled S3/Cloud storage from the admin UI (and confirmed in the diagnosis).
Then regenerated thumbnails. This should have made them local again, right ?

Well… now I have thumbnails in the admin media page for the S3 media uploaded recently. I have some for the previous (non S3) files. But the final image is still not there, S3 or not… and not for all images, some are shown correctly !

---

edit: Apparently the thumbnails generation raised again a "good old" issue https://github.com/YunoHost-Apps/pixelfed_ynh/issues/223 or https://github.com/pixelfed/pixelfed/issues/4275
With file rights such as :
```
.rw-r--r--  27k pixelfed www-data  5 Feb 15:36 randomName.jpg
.rw-r--r-- 9.1k pixelfed pixelfed  5 Feb 16:15 randomName_thumb.jpg
```

Switching to pixelfed:www-data and 750 permissions doesn't fix it.

**Comment by mitexleo** [comment]
> > Regenerating thumbnails didn't work. Worse: some old image thumbnails and final picture where no longer working. The URL where in the S3 storage and no longer local to the Pixelfed instance.
> 
> You want a funny story ? I disabled S3/Cloud storage from the admin UI (and confirmed in the diagnosis). Then regenerated thumbnails. This should have made them local again, right ?
> 
> Well… now I have thumbnails in the admin media page for the S3 media uploaded recently. I have some for the previous (non S3) files. But the final image is still not there, S3 or not… and not for all images, some are shown correctly !
> 
> edit: Apparently the thumbnails generation raised again a "good old" issue [YunoHost-Apps/pixelfed_ynh#223](https://github.com/YunoHost-Apps/pixelfed_ynh/issues/223) or #4275 With file rights such as :
> 
> ```
> .rw-r--r--  27k pixelfed www-data  5 Feb 15:36 randomName.jpg
> .rw-r--r-- 9.1k pixelfed pixelfed  5 Feb 16:15 randomName_thumb.jpg
> ```
> 
> Switching to pixelfed:www-data and 750 permissions doesn't fix it.

Set Directory permission to 775 and File Permisson to 664. Let me know if this works.

**Comment by lapineige** [comment]
Unfortunately I restored a backup so it's no longer present. But I will keep that in mind for a future issue…

Anyway, the good thing is that it uses S3 storage indeed. The bad one is the general issue with file not being accessed correctly from S3, or something like that :thinking:
